### PR TITLE
Rewrite in Typescript: Button

### PR DESCRIPTION
**Pull request by zaki-yama**
ref. #289

### CODE REVIEW

File: src/scripts/Button.tsx [hunk 1]
```diff
@@ -1,25 +1,73 @@
...
-  constructor() {
-    super();
+type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;
+
+export type ButtonType =
+  | 'neutral'
+  | 'brand'
+  | 'destructive'
+  | 'inverse'
+  | 'icon'
```
Comment (zaki-yama) [inline src/scripts/Button.tsx]: Added a prop that were missing in PropTypes, according to story and 
https://archive-2_2_2.lightningdesignsystem.com/components/button-icons

File: src/scripts/Button.tsx [hunk 2]
```diff
@@ -1,25 +1,73 @@
...
+export type ButtonSize = 'x-small' | 'small' | 'medium' | 'large';
+export type ButtonIconSize = typeof ICON_SIZES[number];
+export type ButtonIconAlign = typeof ICON_ALIGNS[number];
+export type ButtonIconMoreSize = 'x-small' | 'small' | 'medium' | 'large';
+
+export type ButtonProps = {
+  className?: string;
+  label?: ReactNode;
+  alt?: string;
+  type?: ButtonType;
```
Comment (zaki-yama) [inline src/scripts/Button.tsx]: @stomita 
When we discussed, we said `type` should be a required prop.
But according to story here: https://github.com/mashmatrix/react-lightning-design-system/blob/master/stories/ButtonStories.js#L73-L75

and LDS's document: https://archive-2_2_2.lightningdesignsystem.com/components/buttons#flavor-base-button

Button with no `type` props should be acceptable. So I left it as it was.

**Comment by zaki-yama** [comment]
TODO: some props (all except `select`) are not displayed in knobs

SS_2019-05-16_0_52_45

ref. current version (https://mashmatrix.github.io/react-lightning-design-system/?knob-notificationText=This%20is%20notification%20text.&knob-type=neutral&knob-disabled=false&knob-category=standard&knob-iconAlign=left&knob-required=undefined&knob-alertTexture=true&knob-size=small&knob-readOnly=undefined&knob-label=Button&selectedKind=Button&selectedStory=Controlled%20with%20knobs&full=0&down=1&left=1&panelRight=0&downPanel=storybooks%2Fstorybook-addon-knobs)

SS_2019-05-16_0_52_57

**Comment by zaki-yama** [comment]
@stomita 
Please review.

The problem in Storybook I commented here https://github.com/mashmatrix/react-lightning-design-system/pull/302#issuecomment-492714560, was resolved in 0748af4 .
But I haven't figured out why this happens and the exact reproduction step. I'll research more.

**Comment by zaki-yama** [comment]
@stomita 
> The problem in Storybook I commented here #302 (comment), was resolved in 0748af4 and a0e42fc.
> But I haven't figured out why this happens and the exact reproduction step. I'll research more.

I resolved this.
When updating addon-knobs version from v3.12.7 (used in `master`) to v5.0.5 (used in this branch), the format of `options` of `select()` had changed.
(key and value were exchanged)

ref. addon-knobs's README in v3.12.7
https://github.com/storybooks/storybook/tree/2e1fe20fc631b9e696e3b6b16c1decb935d147ce/addons/knobs#select

addon-knobs's README in v5.0.5
https://github.com/storybooks/storybook/tree/v5.0.5/addons/knobs#select

ButtonStories options contain `{ '': '(none)' }`, which may have caused the problem.
I fixed and restored the order of knobs.

File: src/scripts/Button.tsx [hunk 3]
```diff
@@ -57,22 +109,18 @@ export default class Button extends Component {
...
       'slds-is-selected': selected,
-      [`slds-button--${size}`]: size && !/^icon-/.test(type),
+      [`slds-button--${size}`]: size && !/^icon-/.test(type || ''),
       [`slds-button--icon-${size}`]:
-        /^(x-small|small)$/.test(size) && /^icon-/.test(type),
+        /^(x-small|small)$/.test(size || '') && /^icon-/.test(type || ''),
     });
 
-    delete props.component;
-    delete props.items;
```
Comment (zaki-yama) [inline src/scripts/Button.tsx]: I couldn't understand why some props are deleted explicitly.

Comment (stomita) [reply src/scripts/Button.tsx]: The code was introduced in this commit:
https://github.com/mashmatrix/react-lightning-design-system/commit/27d911e91f33d51449efdc7b127e796bebf8d0fb#diff-7c21d702e582006eed2cd9e25f5ef6a2

I'm actually not sure why it is needed, but it says React reported warning in this.

**Pull request merged**